extensions/kinesis/src/main/java/com/hazelcast/jet/kinesis/impl/KinesisSinkP.java

```diff
@@ -0,0 +1,502 @@
+            return;
+        }
+
+        List<PutRecordsRequestEntry> entries = buffer.content();
+        sendResult = helper.putRecordsAsync(entries);
+        nextSendTime = currentTime;
+    }
+
+    private void checkIfSendingFinished() {
+        if (sendResult.isDone()) {
+            PutRecordsResult result;
+            try {
+                result = helper.readResult(this.sendResult);
+            } catch (ProvisionedThroughputExceededException pte) {
+                dealWithThroughputExceeded("Data throughput rate exceeded. Backing off and retrying in %d ms");
+                return;
+            } catch (SdkClientException sce) {
+                dealWithSendFailure(sce);
+                return;
+            } catch (Throwable t) {
+                throw rethrow(t);
+            } finally {
+                sendResult = null;
+            }
+
+            pruneSentFromBuffer(result);
+            if (result.getFailedRecordCount() > 0) {
+                dealWithThroughputExceeded("Failed to send " + result.getFailedRecordCount() + " (out of " +
+                        result.getRecords().size() + ") record(s) to stream '" + stream +
+                        "'. Sending will be retried in %d ms, message reordering is likely.");
+            } else {
+                long sleepTimeNanos = throughputController.markSuccess();
+                this.nextSendTime += sleepTimeNanos;
+                this.sleepMetric.set(NANOSECONDS.toMillis(sleepTimeNanos));
+                sendRetryTracker.reset();
+            }
+        }
+    }
+
+    private void dealWithSendFailure(@Nonnull Exception failure) {
+        sendRetryTracker.attemptFailed();
+        if (sendRetryTracker.shouldTryAgain()) {
+            long timeoutMillis = sendRetryTracker.getNextWaitTimeMillis();
+            logger.warning(String.format("Failed to send records, will retry in %d ms. Cause: %s",
+                    timeoutMillis, failure.getMessage()));
+            nextSendTime = System.nanoTime() + MILLISECONDS.toNanos(timeoutMillis);
+        } else {
+            throw rethrow(failure);
+        }
+
+    }
+
+    private void dealWithThroughputExceeded(@Nonnull String message) {
+        long sleepTimeNanos = throughputController.markFailure();
+        this.nextSendTime += sleepTimeNanos;
+        this.sleepMetric.set(NANOSECONDS.toMillis(sleepTimeNanos));
+        logger.warning(String.format(message, NANOSECONDS.toMillis(sleepTimeNanos)));
+    }
+
+    private void pruneSentFromBuffer(@Nullable PutRecordsResult result) {
+        if (result == null) {
+            return;
+        }
+
+        List<PutRecordsResultEntry> resultEntries = result.getRecords();
+        if (result.getFailedRecordCount() > 0) {
+            for (int i = resultEntries.size() - 1; i >= 0; i--) {
+                PutRecordsResultEntry resultEntry = resultEntries.get(i);
+                if (resultEntry.getErrorCode() == null) {
+                    buffer.remove(i);
```

Comment:
Yeah, good points. Had a version similar to this, but have removed it after some comments from Viliam. But will take another look, this looks correct to me now.

Reply:
After looking through it I've realized that this current, convoluted solution came to be because, after making the send window of variable size, it was possible to have send results of a size different than how many items were actually in the buffer, so the logic needed to be more complicated. However it's not the case now, in the final version, so yes, we can make the changes you've suggested and they are much more efficient.